File: 4.8-bc/ts/stage-2-feature-first/src/features/commerce/handler.ts
import type { FastifyInstance } from "fastify";

import { writeError } from "../../shared/httputil.js";
import type { Service } from "./service.js";

interface PlaceBody {
  user_id?: string;
  total_cents?: number;
}

const UUID_RE =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/i;

export class Handler {
  constructor(private readonly svc: Service) {}

  register(app: FastifyInstance): void {
    app.post<{ Body: PlaceBody }>("/orders", async (req, reply) => {
      const body = req.body ?? {};
      if (!body.user_id || !UUID_RE.test(body.user_id)) {
        return reply.code(400).send({ error: "invalid_user_id" });
      }
      if (typeof body.total_cents !== "number") {
        return reply.code(400).send({ error: "invalid_request" });
      }
      try {
        const o = await this.svc.place(body.user_id, body.total_cents);
        return reply.code(201).send({ order_id: o.id, status: o.status });
      } catch (err) {
        req.log.error({ err }, "place order failed");
        return writeError(reply, err);
      }
    });
  }
}

File: 4.8-bc/py/stage-2-feature-first/app/features/commerce/handler.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel
from sqlalchemy.ext.asyncio import AsyncSession

from app.features.commerce.repository import Repository
from app.features.commerce.service import Service
from app.shared.db import get_session

router = APIRouter(tags=["commerce"])


class PlaceReq(BaseModel):
    user_id: str
    total_cents: int


class PlaceResp(BaseModel):
    order_id: str
    status: str


@router.post("/orders", status_code=status.HTTP_201_CREATED, response_model=PlaceResp)
async def place_order(
    req: PlaceReq, session: AsyncSession = Depends(get_session)
) -> PlaceResp:
    try:
        uid = UUID(req.user_id)
    except (ValueError, TypeError) as exc:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail="invalid_user_id"
        ) from exc
    svc = Service(Repository(session))
    o = await svc.place(uid, req.total_cents)
    return PlaceResp(order_id=str(o.id), status=o.status)
